internal/server/admin/key_test.go

```diff
@@ -0,0 +1,228 @@
+package admin
+
+import (
+	"fmt"
+	"net/http"
+	"testing"
+
+	"github.com/gavv/httpexpect/v2"
+	"github.com/google/uuid"
+	v1 "github.com/kong/koko/internal/gen/grpc/kong/admin/model/v1"
+	"github.com/stretchr/testify/require"
+)
+
+func goodKey() *v1.Key {
+	return &v1.Key{
+		Id:   uuid.NewString(),
+		Jwk:  &v1.JwkKey{Kid: uuid.NewString()},
+		Name: "simpleKey-" + uuid.NewString(),
+	}
+}
+
+func validateGoodKey(body *httpexpect.Object) {
+	body.ContainsKey("id")
+	body.ContainsKey("created_at")
+	body.ContainsKey("updated_at")
+	body.ContainsKey("jwk")
+	body.Path("$.jwk").Object().ContainsKey("kid")
+}
+
+func TestKeyCreate(t *testing.T) {
+	s, cleanup := setup(t)
+	defer cleanup()
+	c := httpexpect.New(t, s.URL)
+	t.Run("creates a valid key", func(_ *testing.T) {
+		res := c.POST("/v1/keys").WithJSON(goodKey()).Expect()
+		res.Status(http.StatusCreated)
+		res.Header("grpc-metadata-koko-status-code").Empty()
+		body := res.JSON().Path("$.item").Object()
+		validateGoodKey(body)
+	})
+	t.Run("recreating the same key fails", func(_ *testing.T) {
+		k := goodKey()
+		c.POST("/v1/keys").WithJSON(k).Expect().Status(http.StatusCreated)
+		c.POST("/v1/keys").WithJSON(k).Expect().Status(http.StatusBadRequest)
+	})
+}
+
+func TestKeyUpsert(t *testing.T) {
+	s, cleanup := setup(t)
+	defer cleanup()
+	c := httpexpect.New(t, s.URL)
+	t.Run("upsert a valid key", func(_ *testing.T) {
+		res := c.PUT("/v1/keys/" + uuid.NewString()).
+			WithJSON(goodKey()).
+			Expect()
+		res.Status(http.StatusOK)
+		body := res.JSON().Path("$.item").Object()
+		validateGoodKey(body)
+	})
+	t.Run("re-upserting the same key with different id fails",
+		func(_ *testing.T) {
+			k := goodKey()
+			c.PUT("/v1/keys/" + k.Id).WithJSON(k).Expect().Status(http.StatusOK)
```

Comment:
Something I recently learned about `httpexpect`, it supports placeholders:
```suggestion
			c.PUT("/v1/keys/{id}", k.Id).WithJSON(k).Expect().Status(http.StatusOK)
```

Reply:
:+1: